### ISSUE TITLE
Port to Android

### ISSUE THREAD

**Issue by McC-ITS** [cnlohr/cnping#19]
Hay ;)

I really like your cnping on PC,

Would it be possible to port it to Android?
It would be really handy to have it mobile for wireless testing.
Unfortunately i do not have the skills to figure it out on my own,

thanks in advance for your help
McC

**Comment by dreua** [comment]
I think it would be like writing a new Andrid app with the same functinality as cnping. There would be little to no reusable code from cnping. (Except for the idea and maybe the calculation of "Avg"/"Std"/"Loss".)

Additionaly one had to check if pinging that fast (i.e. without waiting for the previous ping to return or timeout) is possible on Android. Same goes for the "extra size" function if that should be implemented as well. (It is possible to ping like the standard ping tools as there are already Apps for that.)

@cnlohr: If someone would write an Android app, would you approve of a name like cnping-android or should it have its own name without relation to cnping?

**Comment by dreua** [comment]
I think it would be like writing a new Andrid app with the same functinality as cnping. There would be little to no reusable code from cnping. (Except for the idea and maybe the calculation of "Avg"/"Std"/"Loss".)

Additionaly one had to check if pinging that fast (i.e. without waiting for the previous ping to return or timeout) is possible on Android. Same goes for the "extra size" function if that should be implemented as well. (It is possible to ping like the standard ping tools as there are already Apps for that.)

@cnlohr: If someone wrote an Android app, would you approve of a name like cnping-android or should it have its own name without relation to cnping?

**Comment by cnlohr** [comment]
I would absolutely approve the name, even moreso if it used the same convention of white bars on a blue-or-black background that turn red when they breach some timeout with horizontal average and/or standard of deviation bars.

I am just SO SICK of people trying to use command-line ping tools to debug wireless problems.  Tools like HR-ping help, but they just don't cut it for real-time updates!  Any platform, anywhere.  I don't care if it's me, people just need to have cnping-like tools available and need to know they're there to be used.

**Comment by dreua** [comment]
I have done a little bit of research and it seems that my Android knowledge was outdated and my first statement may be wrong (sorry): 
It may well be possible to reuse quite some C code using Googles Android Native Development Kit (NDK) and one might have to use native code anyway to get the fast pings!

I really like cnping as well as the idea to port it to Android and I am planing to try this: If anyone want's to join me please get in touch!

**Comment by cnlohr** [comment]
I have literally zero experience with it.  I can only support you emotionally :-p.  Also, is there any reason not to make it free?  And not require permissions to all your contacts and camera and to send/receive text messages and phone calls?

**Comment by dreua** [comment]
My Android experience is more than zero, but also very limited so I will just give it a try. Your emotional support is greatly appreciated 😄 
A reason not to make it free would be making money but FOSS is just so much cooler, so I'll donate any work and hope for good karma instead. 
There won't be any unnecessary permissions, I promise! Let's hope we don't need root to get a fast ping...

**Comment by cnlohr** [comment]
Not native, but http://stackoverflow.com/questions/3905358/how-to-ping-external-ip-from-java-android

P.S. If you can get RawDraw to work in the Android NDK that would be super cool!  Though I do need to update CNPing's RawDraw API.  RawDraw has been cookin' along. https://github.com/cnlohr/rawdraw

**Comment by hackrid** [comment]
From the Android terminal:

2|FulmicsROM:/ $ ping -i 0.1 192.168.1.1
PING 192.168.1.1 (192.168.1.1) 56(84) bytes of data.
ping: cannot flood; minimal interval, allowed for user, is 200ms

**Comment by cnlohr** [comment]
That's only because the "ping" program restricts that.

**Comment by dreua** [comment]
@hackrid What phone are you using? On my devices, HTC One S (not rooted) and Fairphone 1 (root by default), there is no limit. `ping -i 0 <IP>` pings like crazy!

**Comment by cnlohr** [comment]
I updated to the newest version of rawdraw.  You guys can feel to knock yourselves out now :-D

**Comment by cnlohr** [comment]
P.S.  I would be willing to put $25 of my own $ toward someone making a cnping-type tool for android (provided it works well)

**Comment by dreua** [comment]
Like on Linux and Windows you need root access to open a socket with `SOCK_RAW` flag so I'm wondering which way to go:

1. Require CNPing-Users on Android to root their phone and grant root access to the app. (This would have the best performance and will most likely work best and be closest to cnping.)
1. Try abusing the system provided ping. (Like using many instances of it. May be unreliable or run into performance issues.)
1. Ping by TCP (or maybe UDP?) instead of ICMP. This would require cooperation of the destination device, but many devices already have a web server listening on port 80.
1. Other ideas? 

(Using Java's `InetAddress#isReachable()`method is most likely [not a good idea](http://stackoverflow.com/questions/9922543/why-does-inetaddress-isreachable-return-false-when-i-can-ping-the-ip-address).)

Which one would you guys prefer?

**Comment by hackrid** [comment]
Hi @dreua,

If I read your last message correctly, there's only one option left how to get a cnping Derivate working on android :) 

Or did I miss a point?

**Comment by dreua** [comment]
The problem is, none of the options is perfect. Number one would be best for my own use, *but* other potential users might not be able or wiling to root their phones. Additionally I'm not sure if running an executable with su will work the same on different devices.

btw: I was able to compile most of the C-Code for Android, run it as root (with su) from an Android App and do the pinging. (Just ping, no graphics, no user-interface.) So I think its doable with root but currently the summer term is taking all my time. I hope to continue development in a few months (around August) when I'm done with my exams.

**Comment by hackrid** [comment]
Great! That's a starting point!
Would you like to share your efforts you made so far with the app?

**Comment by Viknet** [comment]
![img_1584](https://user-images.githubusercontent.com/418420/26909528-10efbfd8-4c0a-11e7-9393-5d6b5640860a.jpg)

**Comment by hackrid** [comment]
:D nice

**Comment by hackrid** [comment]
Does this mean you ported cnlohrs rawdraw to Android?

**Comment by Viknet** [comment]
Not exactly :)
I turned it inside out. Rasterizer (in native code) draws on pixel-buffer from android Bitmap.

**Comment by cnlohr** [comment]
@Viknet that makes me cringe a little inside :-p.  But how can I try this?

**Comment by Viknet** [comment]
It's a little unstable and unfinished at so many levels... And needs rooted android.
If you are brave enough, I can create apk in an hour.

**Comment by Viknet** [comment]
I have not tried to install it manually (only through adb).
https://github.com/Viknet/cnping-android/releases/tag/0.0.1

**Comment by Viknet** [comment]
Thank you for some bad news :)
It is SELinux and I know nothing about it... but I will try to write policies on emulator.

**Comment by hackrid** [comment]
I tried to disable selinux as good as possible. But as far I found out there is not much more than setting it to permissive mode.

Regarding the PIE problem:
This situation changed with the security policyies introduced in and rid 4.1. 

What version so you have on your phone?

**Comment by Viknet** [comment]
I will create PIE build in an hour. It raises minimal android version to 4.1 (target was 4.0.2).

"su root setenforce 0" in adb shell should be enough for Unix Domain Socket.
Maybe local Inet socket will work without bothering SELinux.

My phone is under 4.4.2 and not have SELinux.

**Comment by hackrid** [comment]
I tried to setenforce 0.had no luck

**Comment by Viknet** [comment]
PIE build is ready: https://github.com/Viknet/cnping-android/releases/tag/0.0.1a

**Comment by Viknet** [comment]
If device rooted with SuperSU, new build should work around SELinux: https://github.com/Viknet/cnping-android/releases/tag/0.0.1b

**Comment by Viknet** [comment]
@cnlohr You can try last build: [cnping-android-0.0.2](https://github.com/Viknet/cnping-android/releases/tag/0.0.2)

**Comment by cnlohr** [comment]
I don't have a rooted phone, tho.  I guess it's time.

**Comment by cnlohr** [comment]
I know it's been forever, but we've since added support for HTTP Pinging.  It would be very cool if this was something someone did want to work on.

**Comment by dreua** [comment]
Yeah, since Http ping would not require root, that would be the way to go now. Unfortunately I don't have the time to do this... 

@ChristianRomberg  would you like another Android project? ;)

**Comment by cnlohr** [comment]
Now, that I have rawdraw as a first class citizen on Android, and cnping supports httpping, this has become much more tenable.  @dreua would you be interested in chatting on Discord?  I want to get your feelings on if this should be a paid app with more bells and whistles or just a bare bones dead simple app for free.
